internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,668 @@
+package graph
+
+import (
+	"context"
+	"slices"
+
+	"github.com/emirpasic/gods/sets/hashset"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/checkutil"
+	"github.com/openfga/openfga/internal/concurrency"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+const IteratorBatchThreshold = 100
+
+type fastPathSetHandler func(ctx context.Context, iterQueue []*iteratorProducer, iterChan chan *iteratorMsg)
+
+type iteratorMsg struct {
+	iter storage.TupleKeyIterator
+	err  error
+}
+
+type iteratorProducer struct {
+	idx         int
+	iter        storage.TupleKeyIterator
+	initialized bool
+	done        bool
+	producer    chan *iteratorMsg
+}
+
+// NOTE: caller should consider running this in a goroutine to not block.
+func cleanupIteratorProducers(iterProducers []*iteratorProducer) {
+	for _, iterProducer := range iterProducers {
+		if iterProducer.iter != nil {
+			iterProducer.iter.Stop()
+		}
+		for msg := range iterProducer.producer {
+			if msg.iter != nil {
+				msg.iter.Stop()
+			}
+		}
+	}
+}
+
+func pollIteratorProducers(ctx context.Context, iterProducers []*iteratorProducer) ([]*iteratorProducer, error) {
+	for _, producer := range iterProducers {
+		if producer.iter != nil || (producer.done && producer.iter == nil) {
```

Comment:
Can you explain why we need producer.iter == nil when producer.done?  Wouldn't if producer.done == true, line 60 will always be true?